app/src/main/java/de/danoeh/antennapod/activity/CastEnabledActivity.java

```diff
@@ -0,0 +1,135 @@
+package de.danoeh.antennapod.activity;
+
+import android.content.SharedPreferences;
+import android.media.AudioManager;
+import android.os.Bundle;
+import android.preference.PreferenceManager;
+import android.support.v7.app.AppCompatActivity;
+import android.util.Log;
+import android.view.Menu;
+
+import com.google.android.gms.cast.ApplicationMetadata;
+
+import de.danoeh.antennapod.R;
+import de.danoeh.antennapod.core.cast.CastConsumer;
+import de.danoeh.antennapod.core.cast.CastConsumerImpl;
+import de.danoeh.antennapod.core.cast.CastManager;
+import de.danoeh.antennapod.core.cast.SwitchableMediaRouteActionProvider;
+import de.danoeh.antennapod.core.preferences.UserPreferences;
+import de.danoeh.antennapod.core.service.playback.PlaybackService;
+
+/**
+ * Activity that allows for showing the MediaRouter button whenever there's a cast device in the
+ * network.
+ */
+public abstract class CastEnabledActivity extends AppCompatActivity
+        implements SharedPreferences.OnSharedPreferenceChangeListener {
+    public static final String TAG = "CastEnabledActivity";
+
+    protected CastManager mCastManager;
+    private volatile int castUICounter;
+    protected SwitchableMediaRouteActionProvider mMediaRouteActionProvider;
+    protected volatile boolean isCastEnabled;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        PreferenceManager.getDefaultSharedPreferences(getApplicationContext()).
+                registerOnSharedPreferenceChangeListener(this);
+
+        castUICounter = 0;
+        mCastManager = CastManager.getInstance();
+        mCastManager.addCastConsumer(castConsumer);
+        isCastEnabled = UserPreferences.isCastEnabled();
+        onCastConnectionChanged(mCastManager.isConnected());
+    }
+
+    @Override
+    protected void onDestroy() {
+        PreferenceManager.getDefaultSharedPreferences(getApplicationContext())
+                .unregisterOnSharedPreferenceChangeListener(this);
+        mCastManager.removeCastConsumer(castConsumer);
+        super.onDestroy();
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        getMenuInflater().inflate(R.menu.cast_enabled, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onPrepareOptionsMenu(Menu menu) {
+        super.onPrepareOptionsMenu(menu);
+        mMediaRouteActionProvider = mCastManager
+                .addMediaRouterButton(menu.findItem(R.id.media_route_menu_item));
+        mMediaRouteActionProvider.setEnabled(isCastEnabled);
+        return true;
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        castUICounter++;
```

Comment:
Idea is that `CastManager` keeps track of how many active UI elements access it, and the normal practice is to increment on resume and decrement on pause. However, we'd like to not increment/decrement whenever the cast feature is disable. I know that in practice I can't find a use case where the preference is changed while the activity is on resumed state, but it would be good to, just in case, not make many assumptions about it. That's why whenever the preference changes I'd like to know whether the activity is resumed or not. I know I could have used a boolean just for that, but I'm instead using a counter whose values mean 0-paused, 1-resumed, 2-resumed and cast enabled. This is far from being pretty and, as you said, using a volatile doesn't really make it thread safe, although the only problem I see with it in this particular case is that incrementing/decrementing are not atomic operations (i.e., I'm not worried about other effects).
I should redo this part later today, to make it safer and more intuitive.